### PR TITLE
Unify API

### PR DESCRIPTION
**Pull request by jobsorrow**
add support for various printer type.
1. Star Micronics Printer,
2. Bluetooth printer
3. 58 and 80mm paper size
4. dot matrix and thermal printer

Unify printing API with widget printing. and still provide access to deprecated text command printing.

### CODE REVIEW

File: lib/src/printer.dart [hunk 1]
```diff
@@ -25,102 +37,467 @@ class NetworkPrinter {
...
+  }
+
+  static Future<List<dynamic>> discover({bool onlyKnownProfile = true}) async {
+    final escPosPrinterList = await EscPosPrinter.discover();
+    List<StarPortInfo> starPosPrinterList = [];
+
+    try {
+      starPosPrinterList += await StarPrnt.portDiscovery(StarPortType.All);
+    } on PlatformException {
+      /// thrown because no USB device is connected.
```
Comment (Teeraphat-Kongsompree) [inline lib/src/printer.dart:244]: implement error throw

**Review by Teeraphat-Kongsompree — APPROVED** [review]
LGTM

File: lib/src/printer.dart [hunk 2]
```diff
@@ -25,102 +37,509 @@ class NetworkPrinter {
...
+  static String _emulationFor(String modelName) {
+    String emulation = 'StarGraphic';
+    if (modelName != '') {
+      final em = StarMicronicsUtilities.detectEmulation(modelName: modelName);
+      emulation = em!.emulation!;
+    }
+    return emulation;
+  }
+
+  static String printerType(dynamic printer) {
```
Comment (copsterr) [inline lib/src/printer.dart:219]: use verb as a function name might be better.
something like `getPrinterType`

File: lib/src/printer.dart [hunk 3]
```diff
@@ -25,102 +37,509 @@ class NetworkPrinter {
...
+    List<StarPortInfo> starPosPrinterList = [];
+    if (interface == PosPrinterInterface.All ||
+        interface == PosPrinterInterface.Bluetooth) {
+      try {
+        starPosPrinterList +=
+            await StarPrnt.portDiscovery(StarPortType.Bluetooth);
+      } catch (garbageException) {
+        // bluetooth error, continue to discover other interface.
+        throw BluetoothException(
+            'cannot detect bluetooth interface, try to notify user to turn on bluetooth or don\'t detect bluetooth next time');
```
Comment (copsterr) [inline lib/src/printer.dart:262]: Why catch and rethrow?

File: lib/src/printer.dart [hunk 4]
```diff
@@ -25,102 +37,509 @@ class NetworkPrinter {
...
+    }
+  }
+
+  static Future<List<dynamic>> discover(
+      {bool onlyKnownProfile = true,
+      PosPrinterInterface interface = PosPrinterInterface.All}) async {
+    final escPosPrinterList =
+        await EscPosPrinter.discover(interface: interface);
+    List<StarPortInfo> starPosPrinterList = [];
+    if (interface == PosPrinterInterface.All ||
```
Comment (copsterr) [inline lib/src/printer.dart:254]: `interface == PosPrinterInterface.All` might not be neccessary

File: lib/src/printer.dart [hunk 5]
```diff
@@ -25,102 +37,509 @@ class NetworkPrinter {
...
+            'cannot detect network interface, try to notify user to turn on wifi or don\'t detect network interface next time');
+      }
+    } else if (interface == PosPrinterInterface.All ||
+        interface == PosPrinterInterface.USB) {
+      try {
+        starPosPrinterList += await StarPrnt.portDiscovery(StarPortType.USB);
+      } catch (garbageException) {
+        // usb error, continue to discover other interface.
+        throw USBException(
+            'cannot detect USB interface, try to notify user to connect USB printer or don\'t detect USB next time');
```
Comment (copsterr) [inline lib/src/printer.dart:280]: I think `garbageExeception` can be reduced to one block

File: lib/src/printer.dart [hunk 6]
```diff
@@ -25,102 +37,509 @@ class NetworkPrinter {
...
+              'networkPrinterProfile must not be null for NetworkPrinter');
+        }
+        _escPosGenerator = Generator(
+            await CapabilityProfile.load(name: networkPrinterProfile));
+      } else if (_connectedPrinter is BluetoothDevice) {
+        _escPosGenerator = Generator(
+            await CapabilityProfile.load(name: _connectedPrinter.name!));
+      } else {
+        throw Exception('connectedPrinter is invalid printer');
+      }
```
Comment (copsterr) [inline lib/src/printer.dart:343]: you can use switch-case for clarity

File: lib/src/printer.dart [hunk 7]
```diff
@@ -25,102 +37,509 @@ class NetworkPrinter {
...
+
+      final resized = copyResize(dartImage,
+          width: receiptWidthPx, interpolation: Interpolation.nearest);
+      _roundUpGreyScale(resized);
+      // final encoder = PngEncoder();
+      // final finalimg = encoder.encodeImage(resized); // dont delete this line, use for debugging
+      _escPosCommands += _escPosGenerator!.rawBytes([0x1b, 0x55, 0x01]);
+      _escPosCommands += _escPosGenerator!
+          .image(resized, align: align, highDensity: highDensity);
+      _escPosCommands += _escPosGenerator!.rawBytes([0x1b, 0x55, 0x00]);
```
Comment (copsterr) [inline lib/src/printer.dart:433]: raw bytes should use constant

**Review by copsterr — CHANGES_REQUESTED** [review]

**Review by copsterr — APPROVED** [review]
LGTM

**Review by ekawat — COMMENTED** [review]
looks good

**Pull request merged**